test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala

```diff
@@ -67,174 +32,380 @@ class FormatterTest {
+    assertF(" Infinity", "% (" + conversion, PosInf)
+    assertF("(Infinity)", "%+(" + conversion, NegInf)
+    assertF("(Infinity)", "% (" + conversion, NegInf)
+
+    if (acceptUpperCase) {
+      val upConversion = conversion.toUpper
+      assertF("INFINITY", "%" + upConversion, PosInf)
+      assertF("-INFINITY", "%" + upConversion, NegInf)
+      assertF("NAN", "%" + upConversion, NaN)
+    }
+  }
+
   def expectFormatterThrows[T <: Throwable](exeption: Class[T], format: String,
-      args: AnyRef*): Unit = {
+      args: Any*): T = {
     val fmt = new Formatter()
-    assertThrows(exeption, fmt.format(format, args:_*))
+    expectThrows(exeption,
+        fmt.format(format, args.asInstanceOf[Seq[AnyRef]]: _*))
   }
 
-  // Explicitly define these as `var`'s to avoid any compile-time constant folding
-  var IntMax: Int = Int.MaxValue
-  var IntMin: Int = Int.MinValue
-  var ByteMax: Byte = Byte.MaxValue
-  var ByteMin: Byte = Byte.MinValue
-  var ShortMax: Short = Short.MaxValue
-  var ShortMin: Short = Short.MinValue
-
   @Test def `should_provide_b_conversion`(): Unit = {
     assertF("false", "%b", null)
-    assertF(JString.valueOf(true), "%b", true: JBoolean)
-    assertF(JString.valueOf(false), "%b", false: JBoolean)
+    assertF("true", "%b", true)
+    assertF("false", "%b", false)
     assertF("true", "%b", new HelperClass)
+
+    assertF("  false", "%7b", false)
+    assertF("   true", "%7b", true)
+    assertF("false  ", "%-7b", false)
+    assertF("true   ", "%-7b", true)
+
+    assertF("FALSE", "%B", false)
   }
 
   @Test def `should_provide_h_conversion`(): Unit = {
     val x = new HelperClass
-    assertF(Integer.toHexString(x.hashCode()), "%h", x)
-    assertF(Integer.toHexString(x.hashCode()).toUpperCase(), "%H", x)
+    assertF("f1e2a3", "%h", x)
+    assertF("F1E2A3", "%H", x)
+
+    assertF("  f1e2a3", "%8h", x)
+
     assertF("null", "%h", null)
   }
 
-  @Test def `should_provide_s_conversion`(): Unit = {
-    assertFC("foobar", "%s", 0, -1, -1)
-    assertFC("foobar", "%-10s", FormattableFlags.LEFT_JUSTIFY, 10, -1)
-    assertFC("foobar", "%#-10.2s",
-        FormattableFlags.LEFT_JUSTIFY | FormattableFlags.ALTERNATE, 10, 2)
-    assertFC("foobar", "%#10.2S",
-        FormattableFlags.UPPERCASE | FormattableFlags.ALTERNATE, 10, 2)
+  @Test def sConversionWithNonFormattable(): Unit = {
+    assertF("abcdef", "ab%sef", "cd")
+    assertF("true", "%s", true)
+    assertF("12345", "%s", 12345)
+
+    assertF("ABCDEF", "%S", "aBcdeF")
+
     assertF("     hello", "%10s", "hello")
     assertF("hello     ", "%-10s", "hello")
+
+    assertF("null", "%s", null)
+
     if (!executingInJVMOnJDK6)
       expectFormatterThrows(classOf[Exception], "%#s", "hello")
   }
 
-  @Test def `should_fail_s_conversions_without_width`(): Unit = {
-    // Issue #2246
-    expectFormatterThrows(classOf[MissingFormatWidthException], "%-s", "abc")
+  @Test def sConversionWithFormattable(): Unit = {
+    assertFC("foobar", "%s", 0, -1, -1)
```

Comment:
define `assertFC` as a local method here? It is only used in this test. Maybe even define `FormattableClass` here, also inline `"foobar`" into `assertFC`?